src/read/magic_finder.rs

```diff
@@ -0,0 +1,191 @@
+            }
+
+            /* Move cursor to the next chunk, cover magic at boundary by shifting by needle length. */
+            self.cursor = self
+                .cursor
+                .saturating_add(self.finder.needle().len() as u64)
+                .saturating_sub(self.buffer.len() as u64)
+                .clamp(self.bounds.0, self.bounds.1);
+        }
+
+        Ok(None)
+    }
+}
+
+/// A magic bytes finder with an optimistic guess that is tried before
+/// the inner finder begins searching from end. This enables much faster
+/// lookup in files without appended junk, because the magic bytes will be
+/// found directly.
+///
+/// The guess can be marked as mandatory to produce an error. This is useful
+/// if the ArchiveOffset is known and auto-detection is not desired.
+pub struct OptimisticMagicFinder<'a> {
+    inner: MagicFinder<'a>,
+    initial_guess: Option<(u64, bool)>,
+}
+
+/// This is a temporary restriction, to avoid heap allocation in [`Self::next_back`].
+///
+/// We only use magic bytes of size 4 at the moment.
+const STACK_BUFFER_SIZE: usize = 8;
+
+impl<'a> OptimisticMagicFinder<'a> {
+    /// Create a new empty optimistic magic bytes finder.
+    pub fn new_empty() -> Self {
+        Self {
+            inner: MagicFinder::new(&[], (0, 0)),
+            initial_guess: None,
+        }
+    }
+
+    /// Repurpose the finder for different bytes, bounds and initial guesses.
+    pub fn repurpose(
+        &mut self,
+        magic_bytes: &'a [u8],
+        bounds: (u64, u64),
+        initial_guess: Option<(u64, bool)>,
+    ) -> &mut Self {
+        debug_assert!(magic_bytes.len() <= STACK_BUFFER_SIZE);
+
+        self.inner.repurpose(magic_bytes, bounds);
+        self.initial_guess = initial_guess;
+
+        self
+    }
+
+    /// Equivalent to `next_back`, with an optional initial guess attempted before
+    /// proceeding with reading from the back of the file.
+    pub fn next_back<R: Read + Seek>(&mut self, reader: &mut R) -> ZipResult<Option<u64>> {
+        if let Some((v, mandatory)) = self.initial_guess {
+            reader.seek(SeekFrom::Start(v))?;
+
+            let mut buffer = [0; STACK_BUFFER_SIZE];
+            let buffer = &mut buffer[..self.inner.finder.needle().len()];
+
+            // Attempt to match only if there's enough space for the needle
+            if v.saturating_add(buffer.len() as u64) <= self.inner.bounds.1 {
+                reader.read_exact(buffer)?;
+
+                // If a match is found, yield it.
+                if self.inner.finder.rfind(&buffer).is_some() {
+                    self.initial_guess.take();
+                    reader.seek(SeekFrom::Start(v))?;
+                    return Ok(Some(v));
+                }
+            }
+
+            // If a match is not found, but the initial guess was mandatory, return an error.
+            if mandatory {
+                return Ok(None);
+            }
+        }
+
+        self.inner.next_back(reader)
```

Comment:
Will this keep returning the same result if `next_back` is called again? It looks to me like it will (an off-by-one error).

Reply:
There are three cases:
1. no `initial_guess`, the iteration behaves exactly as the inner `MagicFinder`
2. an `initial_guess` is provided and a match is found. The `initial_guess` it destroyed by calling `take()` on it and the finder is converted to case 1
3. no `initial_guess` is provided and no match is found:
    * in the case of a mandatory guess, the iterator terminates (returns `Ok(None)`)
    * in the case of an optional guess, the initial guess falls through. On the next iteration, it will however attempt to check the initial guess again. I will fix this by calling a `take()` in this case, then it will convert to case 1 without making an additional false lookup per iteration.